include/swift/AST/ArchetypeBuilder.h

```diff
@@ -251,7 +255,13 @@ class ArchetypeBuilder {
                             RequirementSource Source);
 
   /// \brief Add all of a generic signature's parameters and requirements.
-  void addGenericSignature(GenericSignature *sig);
+  ///
+  /// \param protocolRequirementSig If true, computes the protocol requirement
+  /// signature (generic signature containing exactly the requirements
+  /// introduced by a protocol) of \c sig, which is expected to be the generic
+  /// signature of a protocol, i.e. \c <Self where Self : Protocol>.
+  void addGenericSignature(GenericSignature *sig,
```

Comment:
I don't think this is the right API. `addGenericSignature()` is meant to introduce an arbitrary signature, but when `protocolRequirementSig == true`, the constraints on that signature are so specific that they'd be better served by calling them out via a new API. Moreover, a client of `ArchetypeBuilder` can do all of this work itself. For a given protocol `proto`, get the protocol's `Self` interface type as a `GenericTypeParamType *` and call `addGenericParameter()`, then `addRequirement` with a newly-synthesized conformance `Requirement` of `Self: proto` and your newly-created `ProtocolRequirementSignatureSelf` source.

To be really specific, I think this API should be on `ProtocolDecl`:

    GenericSignature *getRequirementSignature();

You can make it perform on-demand computation for now (creating a new `ArchetypeBuilder` and following the above process) while we're just dumping() the results for debugging, then start caching the result on the AST, serializing it, etc.